crates/bindings/src/lib.rs

```diff
@@ -225,92 +225,83 @@ pub fn delete_by_rel(table_id: TableId, relation: &[impl Serialize]) -> Result<u
     })
 }
 
-/// A table iterator which yields values of the `TableType` corresponding to the table.
-type TableTypeTableIter<T> = RawTableIter<TableTypeBufferDeserialize<T>>;
-
 // Get the iterator for this table with an optional filter,
-fn table_iter<T: TableType>(table_id: TableId, filter: Option<spacetimedb_lib::filter::Expr>) -> Result<TableIter<T>> {
-    // Decode the filter, if any.
-    let filter = filter
-        .as_ref()
-        .map(bsatn::to_vec)
-        .transpose()
-        .expect("Couldn't decode the filter query");
-
-    // Create the iterator.
-    let iter = sys::iter(table_id, filter.as_deref())?;
+fn table_iter<T: TableType>(table_id: TableId) -> Result<TableIter<T>> {
+    sys::iter(table_id).map(TableIter::new)
+}
 
-    let deserializer = TableTypeBufferDeserialize::new();
-    Ok(RawTableIter::new(iter, deserializer).into())
+fn table_iter_filtered<T: TableType>(
+    table_id: TableId,
+    filter: &spacetimedb_lib::filter::Expr,
+) -> Result<TableIter<T>> {
+    with_row_buf(|buf| {
+        bsatn::to_writer(buf, filter).expect("Couldn't encode the filter query");
+        sys::iter_filtered(table_id, buf).map(TableIter::new)
+    })
 }
 
-/// A trait for deserializing mulitple items out of a single `BufReader`.
-///
-/// Each `BufReader` holds a number of concatenated serialized objects.
-trait BufferDeserialize {
-    /// The type of the items being deserialized.
-    type Item;
+/// A table iterator which yields values of the `TableType` corresponding to the table.
+pub struct TableIter<T: TableType> {
+    /// The underlying source of our `Buffer`s.
+    inner: Option<RowIter>,
 
-    /// Deserialize one entry from the `reader`, which must not be empty.
-    fn deserialize<'de>(&mut self, reader: impl BufReader<'de>) -> Self::Item;
-}
+    /// The current position in the buffer, from which `deserializer` can read.
+    reader: Cursor<Vec<u8>>,
 
-/// Deserialize bsatn values to a particular `T` where `T: TableType`.
-struct TableTypeBufferDeserialize<T> {
     _marker: PhantomData<T>,
 }
 
-impl<T> TableTypeBufferDeserialize<T> {
-    fn new() -> Self {
-        Self { _marker: PhantomData }
-    }
-}
-
-impl<T: TableType> BufferDeserialize for TableTypeBufferDeserialize<T> {
-    type Item = T;
+const DEFAULT_BUFFER_CAPACITY: usize = 0x20_000;
```

Comment:
It was `0x8_000` before, and I was gonna use that, but I felt it was sorta small if we want to avoid reallocation, and 32 KiB is not very much to preallocate. If you have opinions about it I'm super open to changing it. Maybe `DEFAULT_BUFFER_CAPACITY` should even be larger than `ITER_CHUNK_SIZE`?

Reply:
+1

Reply:
In retrospect. Not sure why I thought 0x20_000 B was 20 KiB lol.